src/view/components/TabManager/index.tsx

```diff
@@ -0,0 +1,108 @@
+import { connect, MapStateToProps, MapDispatchToProps, DispatchProp } from 'react-redux';
+import { Tabs as NativeTabs, Text } from 'native-base';
+
+import { Tab, TabList, ElementTab, ElementTabList, StoreTabList } from '../../../store/modules/tabs/types';
+import tabActions from '../../../store/modules/tabs/actions';
+import { State } from '../../../store/types';
+
+/** TabManager OwnProps */
+interface OwnProps {}
+
+interface StoreProps {
+  activeTab: string;
+}
+
+interface DispatchProps extends DispatchProp<State> {
+  setTabs: (tabs: ElementTabList, activeTab: ElementTab) => void;
+  setActiveTab: (activeTab: ElementTab) => void;
+}
+
+interface OwnState {}
+
+/** TabManager props */
+export type Props = OwnProps & StoreProps & DispatchProps;
+
+function createSimpleComponent(tabs: ElementTabList) {
+
+  /**
+  * Page that renders different component depending on which tab the user is on
+  *
+  * Should not be instantiated directly, only extended. Unable to mark abstract due to react-redux
+  */
+  return class TabManager extends React.Component<Props, OwnState> {
+
+    /************************* Member Variables ************************/
+    private tabs: ElementTabList;
+
+    /************************* Member Functions ************************/
+
+    constructor(props: Props) {
+      super(props);
+
+      // Validation
+      if (!tabs.length) {
+        throw new Error('Cannot create TabNavigator with no tabs');
+      }
+
+      this.tabs = tabs;
+    }
+
+    private getActiveElement() {
+      const tab = this.tabs.find(t => t.name === this.props.activeTab);
+      if (!tab) throw new Error('Active tab is not a valid tab');
+
+      return tab.element();
+    }
+
+    /************************* React Lifecycle *************************/
+
+    /** React componentDidMount */
+    public componentDidMount() {
+      this.props.setTabs(this.tabs, this.tabs[0]);
+    }
+
+    /** React render */
+    public render(): JSX.Element {
+      return this.getActiveElement();
+    }
+
+  };
+}
+/******************************* Redux *******************************/
+
+/** Convert TabList to ElementTabList */
+export function toTabs(tabList: ElementTabList): StoreTabList {
+  return tabList.map(toTab);
+}
+
+/** Convert ElementTab to Tab */
+export function toTab(tab: ElementTab): Tab {
+  const {
+    element,
+    ...rest,
+  } = tab;
+
+  return rest;
+}
+
+const mapStateToProps: MapStateToProps<StoreProps, OwnProps, State> = (state, ownProps) => {
+  return {
+    activeTab: (state.tabs.activeTab) ? state.tabs.activeTab.name : '',
+  };
+};
+
+const mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps> = (dispatch) => {
+  return {
+    setActiveTab: tab => dispatch(tabActions.setActiveTab(tab)),
+    setTabs: (tabs, activeTab) => dispatch(tabActions.setTabs(toTabs(tabs), activeTab)),
+  };
+};
+
+/** Creation utility for a TabManager */
+export default function createTabManager(tabs: ElementTabList) {
+  return connect(
```

Comment:
So how do we use this tabmanager? Do we extend it still?

Reply:
# TLDR;
So long story short no. This utility will actually create the class for you. You just pass in the tabs that you want your component to offer, and the function will return a class that you export. This would look similar to the following:
```
const spinnerTabs: ElementTabList = [
  {
    element: () => <SpinnerPage />,
    name: 'Test',
  },
  {
    element: () => <SpinnerPage />,
    name: 'Hello',
  },
];

const SpinnerTabManager = createTabManager(spinnerTabs);

export default MyNewTabManager;
```

# Long Story

## Problem 1
I cannot pass in an abstract class into the `react-redux` `connect` function. This has the implication that anyone not understanding how the base class `TabManager` worked would be tempted to instantiate it directly, which Typescript would not catch but would cause runtime errors, as originally the constructor had the following as a definition:
`(props: Props, tabs: ElementTabList) => void`
And there is no way (as far as I know) to pass in arguments to the constructor using `JSX`. 

## Problem 2
It would seem that after extending the redux-connected class that `connect` returns, there was no need to call `super(props, tabs)` in the constructor of the child class. Not doing this would again result in a bunch of runtime errors, so I'm kind of shocked that Typescript does not enforce it. I suspect if it's an issue with types, it is probably from issues within the `react-redux` typings and not Typescript itself. I'm not even sure if extending the class that `connect` returns actually works either tbh.

## Solution
Seeing as a class is just a function with some syntactic sugar, it occurred to me I could simply create a factory function that creates the class for you with simply one argument, the tabs. I know it looks bizarre in comparison to any other language, but I think it's still very true to the nature of Javascript and am happy I stumbled upon this.

PS: sorry for the essay

Reply:
I also need to document this function far better though

Reply:
In my new PR it now looks like this:

```
/**
 * Create a React Component class to manage the navigation between provided tabs
 *
 * The first tab in the 'tabs' list will be rendered on the far left, and the last
 * will be rendered on the far right.
 *
 * @example
 * const ExampleTabManager = createTabManager(exampleTabs);
 * export default ExampleTabManager
 *
 * @param {ElementTabList} [tabs] The tabs this component should render
 * @returns {React.ComponentClass}
 */
export default function createTabManager(tabs: ElementTabList) {
  return connect(
    mapStateToProps,
    mapDispatchToProps,
  )(createSimpleComponent(tabs));
}
```